pkg/controllers/scheduling/scheduling_controller.go

```diff
@@ -0,0 +1,410 @@
+	clusterClient           clusterclient.Interface
+	clusterLister           clusterlisterv1.ManagedClusterLister
+	clusterSetLister        clusterlisterv1alpha1.ManagedClusterSetLister
+	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister
+	placementLister         clusterlisterv1alpha1.PlacementLister
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister
+	rateLimiter             workqueue.RateLimiter
+}
+
+// NewDecisionSchedulingController return an instance of decisionSchedulingController
+func NewDecisionSchedulingController(
+	clusterClient clusterclient.Interface,
+	clusterLister clusterlisterv1.ManagedClusterLister,
+	clusterSetLister clusterlisterv1alpha1.ManagedClusterSetLister,
+	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister,
+	placementLister clusterlisterv1alpha1.PlacementLister,
+	placementDecisionInformer clusterinformerv1alpha1.PlacementDecisionInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	c := decisionSchedulingController{
+		clusterClient:           clusterClient,
+		clusterLister:           clusterLister,
+		clusterSetLister:        clusterSetLister,
+		clusterSetBindingLister: clusterSetBindingLister,
+		placementLister:         placementLister,
+		placementDecisionLister: placementDecisionInformer.Lister(),
+		rateLimiter: workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 120*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, placementDecisionInformer.Informer()).
+		WithSync(c.sync).
+		ToController("DecisionSchedulingController", recorder)
+}
+
+func (c *decisionSchedulingController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		// ignore placementdecision whose key is not in format: namespace/name
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	klog.V(4).Infof("Reconciling placementdecision %q", queueKey)
+	placementDecision, err := c.placementDecisionLister.PlacementDecisions(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		// no work if placementdecision is deleted
+		c.rateLimiter.Forget(queueKey)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// no work if placementdecision is deleting
+	if !placementDecision.DeletionTimestamp.IsZero() {
+		c.rateLimiter.Forget(queueKey)
+		return nil
+	}
+
+	return c.schedule(ctx, syncCtx, placementDecision)
+}
+
+// schedule makes decisions for the unscheduled decisions in the status of PlacementDecision.
+func (c *decisionSchedulingController) schedule(ctx context.Context, syncCtx factory.SyncContext, placementDecision *clusterapiv1alpha1.PlacementDecision) error {
```

Comment:
this seems a little difficult since you logically need to take a slice here since a placement can have multiple decisions, right?